### ISSUE TITLE
Subtitles causes video to crash

### ISSUE THREAD

**Issue by CHJ85** [peak3d/inputstream.adaptive#946]
HI there. How come you guys do not allow Media stream to only have Audio+VIdeo? Because having it set to "all" causes video to crash in some addons such as Discovery+. I believe it has to do with the way that the website handles subtitles, because setting Media to Video or Audio only solves the issue. But obviously I want both video and audio.
Could you guys please push an update and fix this?
Thank you.

**Comment by glennguy** [comment]
I don't have discovery+ - how would you expect me or anyone else to fix the issue with this information alone?

Btw the solution isn't adding a video+audio option but fixing the bug that is causing the crash. Not every user is you and some may want to have subtitles.

A debug log and any available crash dump/ stack trace would be good to start with, because as far as we know you're on Kodi 17 Alpha with inputstream adaptive v1.0.8, running on who knows what OS and device.

**Comment by CHJ85** [comment]
Right. But adding audio+video would fix the issue because the way that the subtitles are embedded is what crashes Kodi, with Discovery+ anyway. Hence why I brought up Discovery+ as an example.
Btw. The debug is what led me to this addon. That's when I started playing around with the Media settings and learned that both video and audio works fine without the subtitles.
Here's my debug:

<removed> - please attach separately

**Comment by glennguy** [comment]
Please see here for debug log instructions: https://kodi.wiki/view/Log_file/Easy . Debug logging needs to be enabled in settings, the log needs to be the ***full*** log file and not just a snippet, and attached separately either as a .txt attachment in GitHub (just drag and drop) or a link to somewhere such as paste.kodi.tv .

There's no guarantee that an audio+video option would solve your issue, you yourself haven't seen this. It's entirely possible that the crash has something to do with video and audio running at the same time. Many such cases before have been this way. Having a proper debug log (and certainly stack trace if you have one) *may* help pinpoint the problem.

Really keen to help but need the correct info :)

**Comment by CHJ85** [comment]
Thank you. I spoke with the dev behind the Discovery+ addon and he told me to please contact you guys as there's nothing he can do. And this happens to other users as well.
Also. The crash only happens when I have Media set to either All or Subtitle+Video.
However. I can think of other use cases where an Audio+Video option will become in handy.
If you hate subtitles and never use them.
Or if you watch anime on Crunchyroll, because they automatically enable subtitles if the anime is not dubbed.
Always having to turn that off is a bit annoying.
I know this last example is a niche use case but still.
Btw. I know that the author of the Discovery+ addon says there's nothing he can do, but could you have a look through it real quick just in case. Only if you want to. No pressure. :)
https://github.com/Dis90/plugin.video.discoveryplus

**Comment by Dis90** [comment]
@CHJ85 Can you just post debug log so devs can take look of this issue. Making option to set media to audio+video is not right fix for the problem.

**Comment by CHJ85** [comment]
[kodi.log](https://github.com/xbmc/inputstream.adaptive/files/8546643/kodi.log)
Here you go!

**Comment by Dis90** [comment]
That is not debug log. Please enable debug logging Settings -> System -> Logging. Also it is possible that log file contains your discovery+ email and session token. So before posting log make sure you edit them out from log.

> 2022-04-23 11:45:21.157 T:7        INFO <general>: Disabled debug logging due to GUI setting. Level 0.
2022-04-23 11:45:21.157 T:7        INFO <general>: Log level changed to "INFO"

**Comment by CHJ85** [comment]
[kodi.log](https://github.com/xbmc/inputstream.adaptive/files/8546718/kodi.log)
Okay thank you.
Btw. You were right. It did include the token.

**Comment by Dis90** [comment]
Thanks for log.

Just little side note that you're not using latest version of discovery+ add-on. You may want to install repository so you get automatic updates https://github.com/Dis90/repository.dis90/raw/matrix/repo/repository.dis90-1.0.1.zip
In order to get updates from repo you need to install add-on from repository.

**Comment by glennguy** [comment]
Thanks @CHJ85, 

@Dis90 a couple of questions as I can see on your add-on repo this is affecting a few people...

1. Are you able to test yourself and see the crashing on Windows? If so can I send you a debug build of IA and then you send me back a crash dump? If not then either:
2. If there are streams that are not DRM/Widevine and that crash (these are preferred, Widevine can get in the way of debugging) then you could send me a zip with: mpd file, and for each stream (audio/video/subtitle) the init segment and first 3 segments. I can then self host these files (and modify mpd or IA to only give me the right resolution stream that's provided) and reproduce/start debugging
3. If crash only affects Widevine, then I could do the same as above but this would rely on being able to contact the license server from where I am - hopefully there's no special headers/cookies/url params needed...

**Comment by CHJ85** [comment]
@Dis90 Thanks. That did not solve the problem at hand.
However, I appreciate the newly implemented login method.
Much easier than hunting down the cookie string.

**Comment by Dis90** [comment]
I was able to test on Windows and crashing happens with both available streams MPD + Widevine and MPD without DRM. Video I tested crashes everytime in same position (3min 14 sec).

Edit. I was able to get Kodi to crash also on my Mac. Here's log file
[kodi.log](https://github.com/xbmc/inputstream.adaptive/files/8547395/kodi.log)

@glennguy Can you tell what files you want me to send you?

**Comment by glennguy** [comment]
[addon-inputstream.adaptive-19.0.3-windows-x86_64.zip](https://github.com/xbmc/inputstream.adaptive/files/8549068/addon-inputstream.adaptive-19.0.3-windows-x86_64.zip)

@Dis90 can you please install this zip (it's a debug build) on your Windows setup and recreate the crash (non-DRM please)? You should get a crashlog/dump file in your AppData\Roaming\Kodi folder which you can then zip up with the kodi log file as well and attach here.
From there I open the dump file in VS and can see and navigate through the stack where the crash happened. Hopefully will be really easy to see.

**Comment by Dis90** [comment]
Now playback doesn't start
`2022-04-24 08:44:40.526 T:8104    ERROR <general>: IAddonInstanceHandler::ADDON::IAddonInstanceHandler::CreateInstance: inputstream.adaptive returned bad status "Permanent failure" during instance creation`

**Comment by glennguy** [comment]
Hmmm let me check on my end...

**Comment by glennguy** [comment]
You're on Kodi 64bit and installed from zip within Kodi? Or just unzipped into folder?

**Comment by Dis90** [comment]
Yep 64bit Kodi and I installed ISA from zip.

**Comment by glennguy** [comment]
Ok I'll rebuild and check installing from zip on my end. Restarting Kodi made no difference?

**Comment by Dis90** [comment]
I restarted Kodi after installing.

**Comment by glennguy** [comment]
I've downloaded the zip that I posted and installed that (to make sure there's no issues with the file) and it installs and plays without problem.

I'm trying to think of why - the zip as you downloaded it is corrupted (please try re-downloading the zip) or there's some antivirus or something else preventing the dll from executing.

Here is the file as I have it installed:
![image](https://user-images.githubusercontent.com/11909320/164958977-9ccbfa09-bd27-4f29-91bc-aee631172215.png)

**Comment by Dis90** [comment]
I tried to re-download zip with different browser and also tried manually place files to inputstream.adaptive folder but still same result. Also disabled all Windows virus scans. I checked MD5 checksum of inputstream.adaptive.dll and it is same.

**Comment by glennguy** [comment]
Wow the universe really doesn't want me to fix this problem 😆 
How about a reboot?

Other than that, if the crash always happens at 3:14 then I might have to get the files up until that point and try locally here. I see if I can get a script to download the files for you... could you please provide the mpd? mpd seems to be protected ~~but segments not~~.

**Comment by Dis90** [comment]
> Wow the universe really doesn't want me to fix this problem 😆
How about a reboot?

Yeah it looks like it :D Reboot didn't help :(

At least on this video crash happens always at 3:14 but if I skip over that video plays fine but I haven't tested how long.
Here's master.mpd from that video
[master.txt](https://github.com/xbmc/inputstream.adaptive/files/8549155/master.txt)
 
Edit. crashes next at 5:18

**Comment by glennguy** [comment]
Ok cool.

Well let's keep moving forward...

One more test to make - set inputstream.adaptive stream selection setting to manual, then using the OSD during playback change the video stream to the lowest resolution (416x234) and see if it still crashes in the same spot or at all.

If that still crashes then we can save some bandwidth :)

Could you give this project a try to download the stream - https://github.com/Viblast/dash-proxy
If you can get it going then I only need up to the point where it crashes worth of segments sent to me (no idea if this video is 10 minutes long or 2 hours) and the particular video stream that crashes

**Comment by Dis90** [comment]
I did little bit testing and Kodi doesn't crash on ISA version 19.0.1 but obviously subtitles doesn't work because PR #857 is not included.

**Comment by Dis90** [comment]
> One more test to make - set inputstream.adaptive stream selection setting to manual, then using the OSD during playback change the video stream to the lowest resolution (416x234) and see if it still crashes in the same spot or at all.

Crashes also on lowest quality.

> Could you give this project a try to download the stream - https://github.com/Viblast/dash-proxy

Unfortunately I can't get it to download anything.
`Traceback (most recent call last):
  File "./dashproxy.py", line 289, in <module>
    main()
  File "./dashproxy.py", line 286, in main
    run(args)
  File "./dashproxy.py", line 275, in run
    return proxy.run()
  File "./dashproxy.py", line 115, in run
    self.refresh_mpd()
  File "./dashproxy.py", line 124, in refresh_mpd
    logger.log(logging.WARNING, 'Cannot GET the MPD. Server returned %s. Retrying after %ds' % (r.status_code, retry_interval))
NameError: name 'retry_interval' is not defined`

Edit. When I try to go url on browser using incognito mode I get this error:
`Error: Forbidden
Your client does not have permission to get URL /5cf21133-f67e-4eda-86c6-9248d34f03e2/dash_clear_fmp4/x-goog-token=Expires=1650799591 from this server.`

Edit2. Got it working. I had to put url inside "".

**Comment by glennguy** [comment]
> Edit. When I try to go url on browser using incognito mode I get this error:

Yes I get 403 from where I am too - must be a cookie needed?


Ok that tool seems pretty broken.

Right under my nose is youtube-dl and related forks, looking to see now if this will help here (and for other issues in the future)

**Comment by Dis90** [comment]
I actually got it working by putting url inside "". I think url is only valid for 1 hour. I'm now downloading video. Do you have email where I can send link to file after it is complete?

Edit. It is actually valid for 4 hours.

**Comment by glennguy** [comment]
Oh awesome!

To avoid collecting spam I'll explain my email this way: Send it to my github username at kodi.tv

**Comment by Dis90** [comment]
Didn't work. After all video and audio segments was downloaded script failed.
```
Starting a downloader for Representation (period=0 adaptation-set=2 representation=0)
Traceback (most recent call last):
  File "./dashproxy.py", line 289, in <module>
    main()
  File "./dashproxy.py", line 286, in main
    run(args)
  File "./dashproxy.py", line 275, in run
    return proxy.run()
  File "./dashproxy.py", line 115, in run
    self.refresh_mpd()
  File "./dashproxy.py", line 129, in refresh_mpd
    self.handle_mpd(mpd)
  File "./dashproxy.py", line 155, in handle_mpd
    self.ensure_downloader(mpd, rep_addr)
  File "./dashproxy.py", line 173, in ensure_downloader
    downloader.handle_mpd(mpd, self.get_base_url(mpd))
  File "./dashproxy.py", line 211, in handle_mpd
    segments = copy.deepcopy(segment_timeline.findall('mpd:S', ns))
AttributeError: 'NoneType' object has no attribute 'findall'
```

Do you want audio and video segments and if I download few subtitle files by hand?

Edit. I will now try with this version https://github.com/elv-peter/dash-proxy

**Comment by glennguy** [comment]
Thanks for your efforts in this.
If the fork doesn't work then yes the a/v plus a few sub segments would be great. I would need most likely the first 2 subs segs, then the one where it crashes and a couple before it (in the language you're getting). Luckily the sub segments are mostly 30 seconds long. Still 87 of them but at least not 1300 like a/v ones!

**Comment by Dis90** [comment]
Second script also failed to download subtitles. So I downloaded few files manually. I sent link to zip file to your email. Let me know if there's anything else you need :)

**Comment by glennguy** [comment]
Awesome, thanks for that. Downloading now

**Comment by glennguy** [comment]
@Dis90 I think I found the problem, test builds (hopefully this will install and open nicely on your system!) will be here in a couple of minutes: https://jenkins.kodi.tv/blue/organizations/jenkins/xbmc%2Finputstream.adaptive/detail/PR-947/1/artifacts

**Comment by Dis90** [comment]
@glennguy I tested on Mac OS and seems to be working nicely and no crashing anymore. Many thanks :)

**Comment by glennguy** [comment]
Fixed in #947 
Included in release 19.0.4
@CHJ85:
> Or if you watch anime on Crunchyroll, because they automatically enable subtitles if the anime is not dubbed.
Always having to turn that off is a bit annoying.

Open subtitle settings from the OSD while watching a video:

![image](https://user-images.githubusercontent.com/11909320/165000263-5792667a-59d5-4524-a4eb-2bbee9bc91a2.png)

**Comment by CHJ85** [comment]
Thanks for all your hard work.
